contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,728 @@
+   rescheduling the Pod.
+3. `"Off"`: VPA never changes Pod resources. The recommender still sets the
+   recommended resources in the VPA object. This can be used for a “dry run”.
+
+To disable VPA updates the user can do any of the following: (1) change the
+updatePolicy to `"Off"` or (2) delete the VPA or (3) change the Pod labels to no
+longer match the VPA selector.
+
+Note: disabling VPA prevents it from doing further changes, but does not revert
+resources of the running Pods, until they are updated.
+For example, when running a Deployment, the user would need to perform an update
+to revert Pod to originally specified resources.
+
+#### Resource Policy ####
+The resources policy controls how VPA computes the recommended resources.
+In MVP it consists of (optional) lower and upper bound on the request of each
+container.
+The resources policy could later be extended with additional knobs to let the
+user tune the recommendation algorithm to their specific use-case.
+
+#### Recommendation ####
+The VPA resource has an output-only field keeping a recent recommendation,
+filled by the Recommender. This field can be used to obtain a recent
+recommendation even during a temporary unavailability of the Recommender.
+The recommendation consists of the recommended target amount of resources as
+well as an range (min..max), which can be used by the Updater to make decisions
+on when to update the pod.
+In the case of a resource crunch the Updater may decide to squeeze pod resources
+towards the recommended minimum.
+The width of the (min..max) range also reflects the confidence of a
+recommendation. For example, for a workload with a very spiky usage it is much
+harder to determine the optimal balance between performance and resource
+utilization, compared to a workload with stable usage.
+
+#### Complete VPA API object definition ####
+
+```go
+// VerticalPodAutoscaler is the configuration for a vertical pod
+// autoscaler, which automatically manages pod resources based on historical and
+// real time resource utilization.
+type VerticalPodAutoscaler struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the behavior of the autoscaler.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	// +optional
+	Spec VerticalPodAutoscalerSpec
+
+	// Current information about the autoscaler.
+	// +optional
+	Status VerticalPodAutoscalerStatus
+}
+
+// VerticalPodAutoscalerSpec is the specification of the behavior of the autoscaler.
+type VerticalPodAutoscalerSpec {
+	// A label query that determines the set of pods controlled by the Autoscaler.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	Selector *metav1.LabelSelector
```

Comment:
In this design, a poor / incorrect recommendation can completely break a set of pods (by setting too low limits) and thus for most standard kube apps would take the service offline (most people won't get clever with selectors and blue-green setups).  I think this should be called out as a con of the current method below.

Reply:
This is something we try to address with the Disruption Budget. The Updater can also have additional heuristics to not take down pods if more than x% of the group is unhealthy.
Is there anything else that can be done? How is it related to the API design?